README.md
# agh-oop-vending-machine

University assignment for OOP-design-related classes. Fun prohibited. Quit immediately.


changes.md
Zmiany względem oryginalnego projektu:

* zmiana funkcj `wybierzPrzyciskDoWcisniecia` na `wcisnijPrzycisk` oraz
  `wybierProduktDoUzupelnienia` na `uzupelnijProdukt`
* zmiana nazw na angielski
* dodanie funckji `empty()` do `CoinSlot`
* usunięcie funckji `useButton` z klasy `Button`


src/main/agh/cs/lab/vending_machine/Machine.java
package agh.cs.lab.vending_machine;

import java.util.ArrayList;
import java.util.List;

public class Machine {
    CoinSlot slot;

    List<Product> products;
    List<Button> buttons;

    Product productToReceive;

    public Machine(List<Product> products, CoinSlot slot) {
        this.products = products;
        this.slot = slot;

        int idx = 0;
        buttons = new ArrayList<>();
        for (Product p : products) {
            buttons.add(new Button(idx++, p.id));
        }

        productToReceive = null;
    }

    boolean pressButton(int id) {
        if (id >= buttons.size()) {
            System.out.println("No such button!");
            return false;
        }
        Button b = buttons.get(id);
        Product p = products.get(b.getProductId());
        if (productToReceive == null && p.price <= slot.currentValue) {
            if(!p.sell())
                return false;
            productToReceive = p;
            slot.empty();
            return true;
        } else {
            System.out.println("Item not received or not enough money in the slot");
            return false;
        }
    }

    void fillProduct(int id) {
        if (id < products.size())
            products.get(id).fill();
    }

    boolean insertCoin(double value) {
        if (products.isEmpty()) {
            System.out.println("No products in vending machine");
            return false;
        } else {
            return slot.insertCoin(new Coin(value));
        }
    }

    boolean takeProduct() {
        if (productToReceive == null) {
            System.out.println("No product to receive");
            return false;
        } else {
            productToReceive = null;
            return true;
        }
    }

    boolean returnCoins() {
        return slot.returnCoins();
    }

}


src/main/agh/cs/lab/vending_machine/CoinSlot.java
package agh.cs.lab.vending_machine;

import java.util.ArrayList;
import java.util.List;

public class CoinSlot {
    List<Coin> coinsList;

    double currentValue;

    final int coinsLimit;

    boolean filled;

    public CoinSlot(int coinsLimit) {
        this.coinsLimit = coinsLimit;
        this.coinsList = new ArrayList<>(coinsLimit);
        this.currentValue = 0;
        filled = coinsLimit <= 0;
    }

    boolean insertCoin(Coin coin) {
        if (filled) {
            System.out.println("The slot is filled");
            return false;
        } else {
            coinsList.add(coin);
            currentValue += coin.value;
            filled = coinsList.size() == coinsLimit;
            return true;
        }
    }

    boolean returnCoins() {
        if (coinsList.isEmpty()) {
            System.out.println("The slot is empty");
            return false;
        } else {
            empty();
            return true;
        }
    }

    void empty() {
        coinsList.clear();
        currentValue = 0;
    }
}


src/main/agh/cs/lab/vending_machine/VendingMachineREPL.java
package agh.cs.lab.vending_machine;


import java.util.*;

public class VendingMachineREPL {
    static Machine machine;

    public static void main(String[] args) {

        int productCount = 6;
        List<Product> products = new ArrayList<>(productCount);

        Random rng = new Random();
        Set<Integer> generated = new LinkedHashSet<>();
        while (generated.size() < productCount) {
            Integer next = rng.nextInt(2137) + 1;
            generated.add(next);
        }
        Iterator<Integer> iterator = generated.iterator();
        products.add(new Product(iterator.next(), "mamba", 8.0, 8));        // 1
        products.add(new Product(iterator.next(), "monsterek", 5.99, 6));   // 2
        products.add(new Product(iterator.next(), "paluszkis", 4.29, 8));   // 3
        products.add(new Product(iterator.next(), "water", 1.25, 10));      // 4
        products.add(new Product(iterator.next(), "tissues", 0.85, 20));    // 5
        products.add(new Product(iterator.next(), "mentoski", 2.34, 15));   // 6

        machine = new Machine(products, new CoinSlot(2));
    }

    void repl() {

    }

}


src/main/agh/cs/lab/vending_machine/VendingMachineApplication.java
package agh.cs.lab.vending_machine;


import java.util.*;
import java.util.function.Supplier;

public class VendingMachineApplication {
    static Machine machine;

    public static void main(String[] args) {

        int productCount = 6;
        List<Product> products = new ArrayList<>(productCount);

        Random rng = new Random();
        Set<Integer> generated = new LinkedHashSet<>();
        while (generated.size() < productCount) {
            Integer next = rng.nextInt(2137) + 1;
            generated.add(next);
        }
        int idx = 0;
        products.add(new Product(idx++, "espresso", 8.0, 8));        // 0
        products.add(new Product(idx++, "latte", 5.99, 6));   // 1
        products.add(new Product(idx++, "cappucino", 4.29, 8));   // 2
        products.add(new Product(idx++, "water", 1.25, 10));      // 3
        products.add(new Product(idx++, "tissues", 0.85, 20));    // 4
        products.add(new Product(idx++, "mentoski", 2.34, 15));   // 5

        machine = new Machine(products, new CoinSlot(2));

        test();
    }

    static void test() {
        printResult(machine.pressButton(0), "Choose espresso (expected failure)");
        printResult(machine.insertCoin(8), "Insert first coin");
        printResult(machine.takeProduct(), "Take product (expected failure)");
        printResult(machine.pressButton(2), "Choose paluszki");
        printResult(machine.takeProduct(), "Take product");

        printResult(machine.insertCoin(2), "Insert first coin");
        printResult(machine.insertCoin(4), "Insert second coin");
        printResult(machine.insertCoin(1), "Insert third coin (expected failure)");

        printResult(machine.pressButton(3), "Choose water");
        printResult(machine.pressButton(3), "Choose water (expected failure)");

    }

    static void printResult(boolean result, String description) {
        if(result) {
            System.out.println("Action " + description + " succesfull");
        } else {
            System.out.println("Action " + description + " failed");
        }
    }

}
